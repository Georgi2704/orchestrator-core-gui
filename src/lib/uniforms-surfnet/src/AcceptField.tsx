/*
 * Copyright 2019-2020 SURF.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
import "./AcceptField.scss";

import CheckBox from "components/CheckBox";
import I18n, { TranslateOptions } from "i18n-js";
import React, { HTMLProps, useReducer } from "react";
import { Override, connectField } from "uniforms";

type AcceptItem = (string | TranslateOptions)[];
type AcceptValue = "SKIPPED" | "ACCEPTED" | "INCOMPLETE";

export type AcceptFieldProps = Override<
    HTMLProps<HTMLDivElement>,
    {
        disabled: boolean;
        id: string;
        name: string;
        onChange(value?: AcceptValue): void;
        value?: AcceptValue;
        error?: boolean;
        showInlineError?: boolean;
        errorMessage?: string;
        data?: AcceptItem[];
    }
>;

interface AcceptState {
    checks: { [index: number]: boolean };
    skip: boolean;
    allChecked: boolean;
}

interface Action {
    field: number;
    type: string;
    value: boolean;
}

function Accept({
    disabled,
    id,
    name,
    onChange,
    value,
    error,
    showInlineError,
    errorMessage,
    data,
    ...props
}: AcceptFieldProps) {
    const legacy = !data;
    const i18nBaseKey = data ? `forms.fields.${name}_accept` : "forms.fields";

    data = data ?? [
        [name, "label", {}],
        [`${name}_info`, "info", {}],
        [name, "checkbox", {}]
    ];

    let [state, dispatch] = useReducer(
        (state: AcceptState, action: Action) => {
            if (action.type === "skip") {
                state.skip = action.value;
            } else {
                state.checks[action.field] = action.value;
            }

            // We intentionally skip optional checkboxes here
            state.allChecked = data!
                .map((entry: AcceptItem, index: number) => [entry, state.checks[index] || false])
                .filter((entry: (boolean | AcceptItem)[]) => (entry[0] as AcceptItem)[1].endsWith("checkbox"))
                .map((entry: (boolean | AcceptItem)[]) => entry[1] as boolean)
                .every((check: boolean) => check);

            onChange(state.skip ? "SKIPPED" : state.allChecked ? "ACCEPTED" : "INCOMPLETE");

            return { ...state };
        },
        { checks: {}, skip: false, allChecked: false }
    );

    return (
        <section id={id} className="accept-field">
            {data.map((entry: any[], index: number) => {
                const label = I18n.t(`${i18nBaseKey}.${entry[0]}`, entry[2] as TranslateOptions);

                switch (entry[1]) {
                    case "label":
                        return (
                            <div key={index}>
                                <label>{label}</label>
                            </div>
                        );
                    case "info":
                        return (
                            <div key={index}>
                                <em>{label}</em>
                            </div>
                        );
                    case "url":
                        return (
                            <div key={index}>
                                <a href={entry[0] as string} target="_blank" rel="noopener noreferrer">
                                    {entry[0]}
                                </a>
                            </div>
                        );
                    case "warning":
                        return (
                            <div key={index}>
                                <label className="warning">{label}</label>
                            </div>
                        );
                    case "skip":
                        return (
                            <CheckBox
                                key={index}
                                name={entry[0]}
                                onChange={(e: React.FormEvent<HTMLInputElement>) => {
                                    const target = e.target as HTMLInputElement;
                                    dispatch({ field: index, type: "skip", value: target.checked });
                                }}
                                value={state.skip}
                                info={label}
                                className={"skip"}
                            />
                        );
                    default:
                        return (
                            <CheckBox
                                key={index}
                                name={entry[0] + (legacy ? "" : index)} // Index needed to allow checkboxes with same name (we can skip this in legacy mode)
                                className={entry[1].startsWith(">") ? "level_2" : undefined}
                                onChange={(e: React.FormEvent<HTMLInputElement>) => {
                                    const target = e.target as HTMLInputElement;

                                    dispatch({ field: index, type: "check", value: target.checked });
                                }}
                                value={state.checks[index]}
                                info={label}
                                disabled={state.skip || disabled}
                            />
                        );
                }
            })}

            {error && showInlineError && (
                <em className="error">
                    <div className="backend-validation">{errorMessage}</div>
                </em>
            )}
        </section>
    );
}

export default connectField(Accept);
