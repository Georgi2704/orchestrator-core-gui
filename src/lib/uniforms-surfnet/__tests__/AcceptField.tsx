import { mount } from "enzyme";
import React from "react";

import AcceptField from "../src/AcceptField";
import createContext from "./_createContext";

const TEST_ACCEPT_DATA = [
    ["label", "label"],
    ["info", "info"],
    ["warning", "warning"],
    ["http://example.com", "url"],
    ["checkbox1", "checkbox"],
    ["label_with_context", "label", { foo: "bar" }],
    ["sub_checkbox", ">checkbox"],
    ["sub_checkbox2", ">checkbox"],
    ["optional_checkbox", "checkbox?"],
    ["skip_checkbox", "skip"]
];

describe("<AcceptField>", () => {
    test("<AcceptField> - legacy renders an input", () => {
        const element = <AcceptField name="x" />;
        const wrapper = mount(element, createContext({ x: { type: String } }));
        expect(wrapper.find("input")).toHaveLength(1);
    });

    test("<AcceptField> - legacy renders a input which correctly reacts on change", () => {
        const onChange = jest.fn();

        const element = <AcceptField name="x" />;
        const wrapper = mount(element, createContext({ x: { type: String } }, { onChange }));

        expect(wrapper.find("input[name='x']")).toHaveLength(1);
        expect(wrapper.find("input[name='x']").simulate("change", { target: { checked: true } })).toBeTruthy();
        expect(onChange).toHaveBeenLastCalledWith("x", "ACCEPTED");
        expect(wrapper.find("input[name='x']").simulate("change", { target: { checked: false } })).toBeTruthy();
        expect(onChange).toHaveBeenLastCalledWith("x", "INCOMPLETE");
    });

    test("<AcceptField> - renders inputs", () => {
        const element = <AcceptField name="x" />;
        const wrapper = mount(
            element,
            createContext({
                x: {
                    type: String,
                    uniforms: {
                        data: TEST_ACCEPT_DATA
                    }
                }
            })
        );
        expect(wrapper.find("input")).toHaveLength(5);
    });
    test("<AcceptField> - renders a input which correctly reacts on change", () => {
        const onChange = jest.fn();

        const element = <AcceptField name="x" />;
        const wrapper = mount(
            element,
            createContext(
                {
                    x: {
                        type: String,
                        uniforms: {
                            data: TEST_ACCEPT_DATA
                        }
                    }
                },
                { onChange }
            )
        );

        expect(wrapper.find("input[name='checkbox1']")).toHaveLength(1);
        expect(wrapper.find("input[name='sub_checkbox']")).toHaveLength(1);
        expect(wrapper.find("input[name='sub_checkbox2']")).toHaveLength(1);
        expect(wrapper.find("input[name='optional_checkbox']")).toHaveLength(1);
        expect(wrapper.find("input[name='skip_checkbox']")).toHaveLength(1);

        expect(wrapper.find("input[name='checkbox1']").simulate("change", { target: { checked: true } })).toBeTruthy();
        expect(onChange).toHaveBeenLastCalledWith("x", "INCOMPLETE");

        expect(
            wrapper.find("input[name='sub_checkbox']").simulate("change", { target: { checked: true } })
        ).toBeTruthy();
        expect(onChange).toHaveBeenLastCalledWith("x", "INCOMPLETE");

        expect(
            wrapper.find("input[name='optional_checkbox']").simulate("change", { target: { checked: true } })
        ).toBeTruthy();
        expect(onChange).toHaveBeenLastCalledWith("x", "INCOMPLETE");

        expect(
            wrapper.find("input[name='sub_checkbox2']").simulate("change", { target: { checked: true } })
        ).toBeTruthy();
        expect(onChange).toHaveBeenLastCalledWith("x", "ACCEPTED");

        expect(
            wrapper.find("input[name='optional_checkbox']").simulate("change", { target: { checked: false } })
        ).toBeTruthy();
        expect(onChange).toHaveBeenLastCalledWith("x", "ACCEPTED");

        expect(
            wrapper.find("input[name='skip_checkbox']").simulate("change", { target: { checked: true } })
        ).toBeTruthy();
        expect(onChange).toHaveBeenLastCalledWith("x", "SKIPPED");

        expect(
            wrapper.find("input[name='skip_checkbox']").simulate("change", { target: { checked: false } })
        ).toBeTruthy();
        expect(onChange).toHaveBeenLastCalledWith("x", "ACCEPTED");

        expect(wrapper.find("input[name='checkbox1']").simulate("change", { target: { checked: false } })).toBeTruthy();
        expect(onChange).toHaveBeenLastCalledWith("x", "INCOMPLETE");
    });
});
