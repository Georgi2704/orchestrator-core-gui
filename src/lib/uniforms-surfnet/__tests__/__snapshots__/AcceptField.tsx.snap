// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AcceptField> <AcceptField> - legacy renders an input 1`] = `
"<AcceptField name=\\"x\\">
  <Accept name=\\"x\\" id=\\"uniforms-0000-0000\\" disabled={false} label=\\"\\" placeholder=\\"\\" showInlineError={false} changed={false} error={null} errorMessage=\\"\\" field={{ type: [Function: String], optional: false, label: 'X' }} fieldType={[Function: String]} fields={[]} onChange={[Function]} value={[undefined]} required={true}>
    <section id=\\"uniforms-0000-0000\\" required={true} className=\\" accept-field\\">
      <label className=\\"euiFormLabel euiFormRow__label\\">
        [missing &quot;en.forms.fields.x&quot; translation]
      </label>
      <EuiText size=\\"m\\">
        <div className=\\"euiText euiText--medium\\">
          [missing &quot;en.forms.fields.x_info&quot; translation]
        </div>
      </EuiText>
      <EuiCheckbox id=\\"x\\" name=\\"x\\" className={[undefined]} onChange={[Function: onChange]} checked={false} label=\\"[missing \\"en.forms.fields.x\\" translation]\\" readOnly={false} disabled={false} indeterminate={false} compressed={false}>
        <div className=\\"euiCheckbox\\">
          <input className=\\"euiCheckbox__input\\" type=\\"checkbox\\" id=\\"x\\" checked={false} onChange={[Function: onChange]} disabled={false} name=\\"x\\" readOnly={false} />
          <div className=\\"euiCheckbox__square\\" />
          <label className=\\"euiCheckbox__label\\" htmlFor=\\"x\\">
            [missing &quot;en.forms.fields.x&quot; translation]
          </label>
        </div>
      </EuiCheckbox>
    </section>
  </Accept>
</AcceptField>"
`;

exports[`<AcceptField> <AcceptField> - renders inputs 1`] = `
"<AcceptField name=\\"x\\">
  <Accept name=\\"x\\" id=\\"uniforms-0000-0002\\" disabled={false} label=\\"\\" placeholder=\\"\\" showInlineError={false} changed={false} error={null} errorMessage=\\"\\" field={{ uniforms: { data: [ [ 'label', 'label' ], [ 'info', 'info' ], [ 'warning', 'warning' ], [ 'margin', 'margin' ], [ 'value', 'value' ], [ 'http://example.com', 'url' ], [ 'checkbox1', 'checkbox' ], [ 'label_with_context', 'label', { foo: 'bar' } ], [ 'sub_checkbox', '>checkbox' ], [ 'sub_checkbox2', '>checkbox' ], [ 'optional_checkbox', 'checkbox?' ], [ 'skip_checkbox', 'skip' ] ] }, type: [Function: String], optional: false, label: 'X' }} fieldType={[Function: String]} fields={[]} onChange={[Function]} value={[undefined]} required={true} data={[ [ 'label', 'label' ], [ 'info', 'info' ], [ 'warning', 'warning' ], [ 'margin', 'margin' ], [ 'value', 'value' ], [ 'http://example.com', 'url' ], [ 'checkbox1', 'checkbox' ], [ 'label_with_context', 'label', { foo: 'bar' } ], [ 'sub_checkbox', '>checkbox' ], [ 'sub_checkbox2', '>checkbox' ], [ 'optional_checkbox', 'checkbox?' ], [ 'skip_checkbox', 'skip' ] ]}>
    <section id=\\"uniforms-0000-0002\\" required={true} className=\\" accept-field\\">
      <label className=\\"euiFormLabel euiFormRow__label\\">
        [missing &quot;en.forms.fields.x_accept.label&quot; translation]
      </label>
      <EuiText size=\\"m\\">
        <div className=\\"euiText euiText--medium\\">
          [missing &quot;en.forms.fields.x_accept.info&quot; translation]
        </div>
      </EuiText>
      <label className=\\"euiFormLabel euiFormRow__label warning\\">
        [missing &quot;en.forms.fields.x_accept.warning&quot; translation]
      </label>
      <br />
      <div>
        <input value=\\"value\\" disabled={true} />
      </div>
      <div>
        <a href=\\"http://example.com\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">
          http://example.com
        </a>
      </div>
      <EuiCheckbox id=\\"checkbox16\\" name=\\"checkbox16\\" className={[undefined]} onChange={[Function: onChange]} checked={false} label=\\"[missing \\"en.forms.fields.x_accept.checkbox1\\" translation]\\" readOnly={false} disabled={false} indeterminate={false} compressed={false}>
        <div className=\\"euiCheckbox\\">
          <input className=\\"euiCheckbox__input\\" type=\\"checkbox\\" id=\\"checkbox16\\" checked={false} onChange={[Function: onChange]} disabled={false} name=\\"checkbox16\\" readOnly={false} />
          <div className=\\"euiCheckbox__square\\" />
          <label className=\\"euiCheckbox__label\\" htmlFor=\\"checkbox16\\">
            [missing &quot;en.forms.fields.x_accept.checkbox1&quot; translation]
          </label>
        </div>
      </EuiCheckbox>
      <label className=\\"euiFormLabel euiFormRow__label\\">
        [missing &quot;en.forms.fields.x_accept.label_with_context&quot; translation]
      </label>
      <EuiCheckbox id=\\"sub_checkbox8\\" name=\\"sub_checkbox8\\" className=\\"level_2\\" onChange={[Function: onChange]} checked={false} label=\\"[missing \\"en.forms.fields.x_accept.sub_checkbox\\" translation]\\" readOnly={false} disabled={false} indeterminate={false} compressed={false}>
        <div className=\\"euiCheckbox level_2\\">
          <input className=\\"euiCheckbox__input\\" type=\\"checkbox\\" id=\\"sub_checkbox8\\" checked={false} onChange={[Function: onChange]} disabled={false} name=\\"sub_checkbox8\\" readOnly={false} />
          <div className=\\"euiCheckbox__square\\" />
          <label className=\\"euiCheckbox__label\\" htmlFor=\\"sub_checkbox8\\">
            [missing &quot;en.forms.fields.x_accept.sub_checkbox&quot; translation]
          </label>
        </div>
      </EuiCheckbox>
      <EuiCheckbox id=\\"sub_checkbox29\\" name=\\"sub_checkbox29\\" className=\\"level_2\\" onChange={[Function: onChange]} checked={false} label=\\"[missing \\"en.forms.fields.x_accept.sub_checkbox2\\" translation]\\" readOnly={false} disabled={false} indeterminate={false} compressed={false}>
        <div className=\\"euiCheckbox level_2\\">
          <input className=\\"euiCheckbox__input\\" type=\\"checkbox\\" id=\\"sub_checkbox29\\" checked={false} onChange={[Function: onChange]} disabled={false} name=\\"sub_checkbox29\\" readOnly={false} />
          <div className=\\"euiCheckbox__square\\" />
          <label className=\\"euiCheckbox__label\\" htmlFor=\\"sub_checkbox29\\">
            [missing &quot;en.forms.fields.x_accept.sub_checkbox2&quot; translation]
          </label>
        </div>
      </EuiCheckbox>
      <EuiCheckbox id=\\"optional_checkbox10\\" name=\\"optional_checkbox10\\" className={[undefined]} onChange={[Function: onChange]} checked={false} label=\\"[missing \\"en.forms.fields.x_accept.optional_checkbox\\" translation]\\" readOnly={false} disabled={false} indeterminate={false} compressed={false}>
        <div className=\\"euiCheckbox\\">
          <input className=\\"euiCheckbox__input\\" type=\\"checkbox\\" id=\\"optional_checkbox10\\" checked={false} onChange={[Function: onChange]} disabled={false} name=\\"optional_checkbox10\\" readOnly={false} />
          <div className=\\"euiCheckbox__square\\" />
          <label className=\\"euiCheckbox__label\\" htmlFor=\\"optional_checkbox10\\">
            [missing &quot;en.forms.fields.x_accept.optional_checkbox&quot; translation]
          </label>
        </div>
      </EuiCheckbox>
      <EuiCheckbox id=\\"skip_checkbox\\" name=\\"skip_checkbox\\" onChange={[Function: onChange]} checked={false} label=\\"[missing \\"en.forms.fields.x_accept.skip_checkbox\\" translation]\\" className=\\"skip\\" disabled={false} indeterminate={false} compressed={false}>
        <div className=\\"euiCheckbox skip\\">
          <input className=\\"euiCheckbox__input\\" type=\\"checkbox\\" id=\\"skip_checkbox\\" checked={false} onChange={[Function: onChange]} disabled={false} name=\\"skip_checkbox\\" />
          <div className=\\"euiCheckbox__square\\" />
          <label className=\\"euiCheckbox__label\\" htmlFor=\\"skip_checkbox\\">
            [missing &quot;en.forms.fields.x_accept.skip_checkbox&quot; translation]
          </label>
        </div>
      </EuiCheckbox>
    </section>
  </Accept>
</AcceptField>"
`;
